### PR TITLE
Deserialize inner functions

### PR DESCRIPTION
**Pull request by noahlemen**
> note that this PR is stacked on https://github.com/mhahnenberg/v8/pull/33

- Adds new uncompiled data type to SFI: UncompiledDataWithInnerBinAstParseData
  - Same as UncompiledDataWithBinAstParseData, but also has offset and length
- Deserializer uses offset to start parsing at inner function
  - Length is not currently used, but we may need it later for skipping inner function deserialization

While deserializing data, we now construct UncompiledDataWithInnerBinAstParseData, set it on the FunctionLiteral, and in the Bytecode Generator set it on the SFI. When deserializing, if an SFI has UncompiledDataWithInnerBinAstParseData, use the BinAst data, offset, and length from it to deserialize the inner function.

During inner function deserialization, I had to do some wacky stuff to resolve the outside variables that were not located contiguously inside the inner function's serialized data. To do this, I now encode the variable reference IDs as offsets within the data instead of an incrementing ID so the deserializer can jump to noncontiguous sections of the data to deserialize the variables. I'm not sure if this is necessary - it seems like ideally we'd use the outer scope to resolve this, but I wasn't able to figure out a way to do this, and afaict the current approach is working?

### Issues
- Test scripts in d8 now fail unless run with `--no-wait-for-wasm`. I wasn't able to figure out why:
```
* thread #1, name = 'd8', stop reason = signal SIGSEGV: invalid address (fault address: 0x120)
  * frame #0: 0x00007ffff5abfd74 libpthread.so.0`__pthread_mutex_lock + 4
    frame #1: 0x00007ffff5f113fe libv8_libbase.so`v8::base::LockNativeHandle(pthread_mutex_t*) at mutex.cc:57:16
    frame #2: 0x00007ffff5f113de libv8_libbase.so`v8::base::Mutex::Lock() at mutex.cc:95:3
    frame #3: 0x00007ffff7994da0 libv8.so`v8::internal::wasm::WasmEngine::HasRunningCompileJob(v8::internal::Isolate*) at wasm-engine.cc:844:20
    frame #4: 0x000055555558e321 d8`v8::platform::MessageLoopBehavior std::__Cr::__function::__policy_invoker<v8::platform::MessageLoopBehavior ()>::__call_impl<std::__Cr::__function::__default_alloc_func<v8::Shell::CompleteMessageLoop(v8::Isolate*)::$_2, v8::platform::MessageLoopBehavior ()> >(std::__Cr::__function::__policy_storage const*) at d8.cc:3362:38
    frame #5: 0x00005555555800b9 d8`v8::(anonymous namespace)::ProcessMessages(v8::Isolate*, std::__Cr::function<v8::platform::MessageLoopBehavior ()> const&) at functional:2203:16
    frame #6: 0x000055555557f7a7 d8`v8::Shell::RunMain(v8::Isolate*, bool) at d8.cc:3368:10
    frame #7: 0x0000555555581cef d8`v8::Shell::Main(int, char**) at d8.cc:3916:18
    frame #8: 0x00007ffff57c9042 libc.so.6`__libc_start_main + 242
    frame #9: 0x0000555555569d7a d8`_start + 42
```
- Deserialization work is duplicated for inner functions. They are first deserialized when the containing function parses, and again when the inner function itself is parsed.

### CODE REVIEW

**Comment by noahlemen** [comment]
fixed issue deserializing `explode2` by removing toplevel condition around offset/length

wasm issue persists, though

File: src/objects/shared-function-info-inl.h [hunk 1]
```diff
@@ -672,6 +690,34 @@ void SharedFunctionInfo::ClearBinAstParseData() {
   DCHECK(!HasUncompiledDataWithBinAstParseData());
 }
 
+void SharedFunctionInfo::ClearInnerBinAstParseData() {
+  DCHECK(HasUncompiledDataWithInnerBinAstParseData());
+  UncompiledDataWithInnerBinAstParseData data = uncompiled_data_with_inner_bin_ast_parse_data();
+
+  // Trim off the pre-parsed scope data from the uncompiled data by swapping the
+  // map, leaving only an uncompiled data without pre-parsed scope.
```
Comment (mhahnenberg) [inline src/objects/shared-function-info-inl.h]: Is this comment accurate?

File: src/parsing/abstract-parser.h [hunk 2]
```diff
@@ -2004,41 +2004,89 @@ void AbstractParser<Impl>::ParseFunction(
...
-
-    RuntimeCallTimerScope runtime_timer(impl()->runtime_call_stats_,
-                                      RuntimeCallCounterId::kDeserializeBinAst);
+  if (V8_UNLIKELY(shared_info->HasUncompiledDataWithBinAstParseData() ||
+                  shared_info->HasUncompiledDataWithInnerBinAstParseData())) {
+    RuntimeCallTimerScope runtime_timer(
+        impl()->runtime_call_stats_, RuntimeCallCounterId::kDeserializeBinAst);
     auto start = std::chrono::high_resolution_clock::now();
+    bool is_inner = shared_info->HasUncompiledDataWithInnerBinAstParseData();
+    if (is_inner) {
```
Comment (mhahnenberg) [inline src/parsing/abstract-parser.h:2016]: Hmm is there any way we could dedupe most of the code in these two branches, maybe through the use of Optional?

File: src/parsing/binast-deserializer-inl.h [hunk 3]
```diff
@@ -268,21 +270,28 @@ inline BinAstDeserializer::DeserializeResult<Variable*> BinAstDeserializer::Dese
...
 
-  return {variable, offset};
+  if (variable_result != variables_by_id_.end()) {
+    return {variable_result->second, offset};
+  } else {
+    return {
+      DeserializeScopeVariable(serialized_binast, variable_reference.value, scope).value,
+      offset
+    };
+  }
```
Comment (mhahnenberg) [inline src/parsing/binast-deserializer-inl.h]: What case are we handling here? When wouldn't the variable be found in variables_by_id_? Might be good to add a comment explaining.

File: src/parsing/binast-serialize-visitor.h [hunk 4]
```diff
@@ -557,6 +557,13 @@ inline void BinAstSerializeVisitor::SerializeDeclarationScope(DeclarationScope*
     scope->needs_private_name_context_chain_recalc_,
   });
 
+  if (scope->has_rest_) {
+    printf(
+        "BinAstSerializeVisitor encountered unsupported rest parameter on "
+        "DeclarationScope\n");
+    encountered_unhandled_node_ = true;
+  }
```
Comment (mhahnenberg) [inline src/parsing/binast-serialize-visitor.h]: Nice find

**Review by mhahnenberg — CHANGES_REQUESTED** [review]
Looking really good! A few random nits/requests.

File: src/objects/shared-function-info-inl.h (hunk 1 above)
Comment (noahlemen) [reply src/objects/shared-function-info-inl.h]: heh, no looks like its not. i copied from ClearBinAstParseData, which looks like it came from being copied from ClearPreparseData? and it seems inaccurate there too... will remove both

Comment (mhahnenberg) [reply src/objects/shared-function-info-inl.h]: Ha, whoops. Copy paste coding strikes again :sweat_smile:

File: src/parsing/abstract-parser.h (hunk 2 above)
Comment (noahlemen) [reply src/parsing/abstract-parser.h:2016]: yeah whoops, meant to rewrite this. after your change with the handles here i had some trouble keeping it not duped, so duplicated it and meant to figure it out later

File: src/parsing/binast-serialize-visitor.h [hunk 5]
```diff
@@ -652,6 +667,11 @@ inline void BinAstSerializeVisitor::VisitFunctionLiteral(FunctionLiteral* functi
     VisitNode(statement);
   }
 
+  int unhandled_inner_nodes = encountered_unhandled_nodes_ - original_unsupported_nodes;
+  if (unhandled_inner_nodes > 0) {
+    skipped_functions_++;
+  } 
+
```
Comment (mhahnenberg) [inline src/parsing/binast-serialize-visitor.h:674]: Noice

File: src/parsing/binast-deserializer-inl.h [hunk 6]
```diff
@@ -275,6 +275,8 @@ BinAstDeserializer::DeserializeVariableReference(uint8_t* serialized_binast,
   if (variable_result != variables_by_id_.end()) {
     return {variable_result->second, offset};
   } else {
+    // When using an offset to deserialize inner functions, we may encounter variable references that are serialized earlier in the data.
+    // To deal with this noncontiguous data, jump there to deserialize the variable, but do not follow the resulting offset.
```
Comment (mhahnenberg) [inline src/parsing/binast-deserializer-inl.h:279]: Aha, now I get it :sweat_smile:

**Review by mhahnenberg — APPROVED** [review]
LGTM!

**Pull request merged**